### PR TITLE
create custom function inet_ntoa,inet_aton[46]

### PR DESCRIPTION
**Pull request by ng-labo**
We have tried to create some custom functions.
This is planned in [ENG-4543].
We assume that VARBINARY(16) is assigned for IPv6 address column .

### CODE REVIEW

**Comment by benjaminballard** [comment]
Hi Ogawa-san,

Thank you for submitting this pull request and for taking the time to learn how to create a custom function.  We are eager to review and merge this.

Before we can do so, we need you to sign a Contributor License Agreement which assigns copyright of the contribution to VoltDB.  The form is on our web site here: https://www.voltdb.com/contributor-license-agreement

You can sign this electronically or send us an image of the signed document.  We will countersign it and return this to you, then we can merge the code.

Thank you,
Ben

**Comment by benjaminballard** [comment]
Hi Ogawa-san,

I forgot to mention, please email the signed agreement to support@voltdb.com, or you could attach it to the pull request if you prefer.

Thanks,
Ben Ballard
bballard@voltdb.com

**Comment by ng-labo** [comment]
I sent the signed agreement  to support@voltdb.com.

**Comment by benjaminballard** [comment]
Thanks, we'll review this now.

**Comment by benjaminballard** [comment]
Hi Naoki,

I sent you the countersigned CLA for your records.

There may be some additional comments or requests here from VoltDB engineers as the code is reviewed.

Thanks,
Ben

**Comment by BillWhite** [comment]
Hello, Ogawa-san.  I'm Bill White with VoltDB, and I work in the SQL team.  Thanks for submitting this.

I'm finishing up a review of your code.  I'll have some comments and some requested changes in a short time.

**Comment by BillWhite** [comment]
Hello, Ogawa-san,

I decided maybe it would be better for me to spend some time working on this issue, rather than have you try to puzzle out what I asked for in my code review.   So, I'm working on this right now on a branch which has your changes pulled in.  I wouldn't want us both to make the same changes and waste our time, so maybe you should hold up and see where I get with it.

If you have any questions or suggestions, please let me know.  Otherwise, just watch this space and I'll let you know when I have something to show you.

Does that sound ok to you?

**Comment by ng-labo** [comment]
I thank you for taking a time to review, and suggestion. I read request change comment and I arrange this code issues. I feel sorry to be late replying. 

 My suggestion for change request is reorganization to 4 custom function. Could you tell me what do you think about definition 4 custom functions?
  I write my idea about the 4 custom functions below.

First, I try to summarize the point for me following 
1) Shortage of document how function should be implemented.
 First I will define custom function again.(of course I'm not going to get much change, I'll get a bit-modify to resolve ambiguous and to get easy to understand.)
2) Function name
 I plan that custom function name make match to MySQL function compatibility with modification for variable type.(thanks for your comment)
3) Adding UnitTest code 
 I try it.

Let me be sure what parameter for 8 function you mentioned in the forth comment, 
I understand that  (IPv4|IPv6) by (direction[a2n|n2a]) by "**TYPE[numeric|binary]**", not byte-ordering.
Functions are , i.e.
inet4_aton ('1.2.3.4' => 0x01020304)
inet4_atob ('1.2.3.4' => [0x01,0x02,0x03,0x04])
inet4_ntoa (0x01020304 => '1.2.3.4')
inet4_btoa ([0x01,0x02,0x03,0x04] => '1.2.3.4')
inet6_aton ..
inet6_atob ..
inet6_ntoa ..
inet6_btoa ..
(I beleave 'b' means binary)(IPv6's numerical type will be assigned to BigDecimal)(ommited v6 example) 
**I'm not going to implement all above.** IPv6's numerical type is not important for networking software. Its form is difficult for bit operation.
So inet6_ntoa,inet6_aton are reserve,also inet4_atob,inet4_btoa is not needed for us.

 In the first pull request, One point is clearly my fault. In the case 'inet_ntoa ' from verbinary(4) to representing string, Its byte ordering is out of my plan. I thank your point out. In fact, We do not need this pattern in our software.
 So I would like to update code using 4-function named **inet4_aton,inet4_ntoa,inet6_atob,inet6_btoa** to make follow your pointing out and to get we need functions.

File: src/ee/expressions/functionexpression.h [hunk 1]
```diff
@@ -261,6 +261,11 @@ static const int FUNC_VOLT_REGEXP_POSITION             = 20041;
 static const int FUNC_VOLT_ROUND                       = 20042;
 static const int FUNC_VOLT_STR                         = 20043;
 
+// our local function
+static const int FUNC_MY_INET_NTOA                     = 20044;
+static const int FUNC_MY_INET_ATON4                    = 20045;
```
Comment (ng-labo) [inline src/ee/expressions/functionexpression.h]: I agree.

File: src/hsqldb19b3/org/hsqldb_voltpatches/FunctionForVoltDB.java [hunk 2]
```diff
@@ -154,6 +154,11 @@ private FunctionId(String name, Type type, int id, int typeParameter, Type[] par
         static final int FUNC_VOLT_ROUND                  = 20042;
         static final int FUNC_VOLT_STR                    = 20043;
 
+        // our local functions for networking
+        static final int FUNC_MY_INET_NTOA                = 20044;
+        static final int FUNC_MY_INET_ATON4               = 20045;
+        static final int FUNC_MY_INET_ATON6               = 20046;
+
```
Comment (ng-labo) [inline src/hsqldb19b3/org/hsqldb_voltpatches/FunctionForVoltDB.java]: I agree.

File: src/hsqldb19b3/org/hsqldb_voltpatches/FunctionForVoltDB.java [hunk 3]
```diff
@@ -378,6 +383,19 @@ private FunctionId(String name, Type type, int id, int typeParameter, Type[] par
...
+                    singleParamList),
+
+            new FunctionId("inet_aton4", Type.SQL_BIGINT, FUNC_MY_INET_ATON4, -1,
+                    new Type[] { Type.SQL_VARCHAR },
+                    singleParamList),
+
+            new FunctionId("inet_aton6", Type.SQL_VARBINARY, FUNC_MY_INET_ATON6, -1,
+                    new Type[] { Type.SQL_VARCHAR },
+                    singleParamList),
+
```
Comment (ng-labo) [inline src/hsqldb19b3/org/hsqldb_voltpatches/FunctionForVoltDB.java]: I agree that being compatible with MySQL's inet function for these custom functions name. Or like MySQL's inet function.

File: src/ee/expressions/bitwisefunctions.h [hunk 4]
```diff
@@ -231,5 +231,83 @@ template<> inline NValue NValue::call<FUNC_VOLT_BIT_SHIFT_RIGHT>(const std::vect
...
+           << ((v&0xff00)>>8) << "." << (v&0xff) ;
+        std::string res (ss.str());
+        return getTempStringValue(res.c_str(), res.length());
+    }
+    if(getValueType() == VALUE_TYPE_VARBINARY){
+        std::string token = toString();
+        std::size_t sz = token.length();
+        if(sz==8){
+            uint32_t v;
+            catalog::Catalog::hexDecodeString(token, (char *) &v);
```
Comment (ng-labo) [inline src/ee/expressions/bitwisefunctions.h]: It is not my plan for the case of from varbinary(4) value to IPv4 string. Its my fault.
It should be passed to inet_ntoa function directory without hton function.

File: src/ee/expressions/bitwisefunctions.h [hunk 5]
```diff
@@ -231,5 +231,83 @@ template<> inline NValue NValue::call<FUNC_VOLT_BIT_SHIFT_RIGHT>(const std::vect
...
+               << ((v&0xff00)>>8) << "." << (v&0xff) ;
+            std::string res (ss.str());
+            return getTempStringValue(res.c_str(),res.length());
+        }else if(sz==32){
+            const size_t IPV6BINLEN = 16;
+            char ipv6bin[IPV6BINLEN];
+            char str[INET6_ADDRSTRLEN];
+
+            catalog::Catalog::hexDecodeString(toString(),ipv6bin);
+            inet_ntop(AF_INET6, ipv6bin, str, INET6_ADDRSTRLEN);
```
Comment (ng-labo) [inline src/ee/expressions/bitwisefunctions.h]: hexDecodeString is not my friend, yes , but I have checked its code and I got to know its procedure is I want.
Method procedure is not problem, but I didn't check whether this namespace is permitted  at sql-function on coding guide line or so.

**Comment by ng-labo** [comment]
I would like to accept your suggestion for last comment.  I keep my posted code, and I check your update.

We hope you keep in mind that our request is 4 custom function, as I wrote above,
- (input) IPv4 Address String  (output) BigInt value representing for IPv4 address.
- (input) IPv6 Address String  (output) VarBinary(16) value representing for IPv6 address.
- (input) BigInt value representing for IPv4 address. (output)  IPv4 Address String.
- (input) VarBinary(16)value representing for IPv6 address. (output)  IPv6 Address String.

**Pull request closed without merging**